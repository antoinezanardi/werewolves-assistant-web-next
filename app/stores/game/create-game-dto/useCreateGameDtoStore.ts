import { defineStore } from "pinia";
import { ref } from "vue";
import { DEFAULT_GAME_OPTIONS } from "~/composables/api/game/constants/game-options/game-options.constants";
import type { CreateGameAdditionalCardDto } from "~/composables/api/game/dto/create-game/create-game-additional-card/create-game-additional-card.dto";

import { CreateGamePlayerDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { GameAdditionalCardRecipientRoleName } from "~/composables/api/game/types/game-additional-card/game-additional-card.types";
import { GameOptions } from "~/composables/api/game/types/game-options/game-options.class";
import { ADDITIONAL_CARDS_DEPENDANT_ROLES } from "~/composables/api/role/constants/role.constants";
import type { RoleName } from "~/composables/api/role/types/role.types";
import { StoreIds } from "~/stores/enums/store.enum";
import { useRolesStore } from "~/stores/role/useRolesStore";
import { useLocalStorage } from "@vueuse/core";
import { LocalStorageKeys } from "~/utils/enums/local-storage.enums";

const useCreateGameDtoStore = defineStore(StoreIds.CREATE_GAME_DTO, () => {
  const rolesStore = useRolesStore();
  const { getRoleWithNameInRoles } = rolesStore;

  const createGameOptionsDtoFromLocalStorage = useLocalStorage(LocalStorageKeys.GAME_OPTIONS, DEFAULT_GAME_OPTIONS, { mergeDefaults: true });

  const defaultCreateGameDto = CreateGameDto.create({
    players: [],
    options: DEFAULT_GAME_OPTIONS,
  });

  const createGameDto = ref<CreateGameDto>(CreateGameDto.create({
    ...defaultCreateGameDto,
    options: createGameOptionsDtoFromLocalStorage.value,
  }));

  const createGameOptionsDto = computed<GameOptions>(() => createGameDto.value.options);

  const doesCreateGameDtoContainPositionDependantRoles = computed<boolean>(() => {
    const positionDependantRoles: RoleName[] = ["rusty-sword-knight", "bear-tamer", "fox"];
    const playersWithPositionDependantRoles = getPlayersWithAnyRoleNameInCreateGameDto(positionDependantRoles);

    return playersWithPositionDependantRoles.length > 0;
  });

  const doesCreateGameDtoContainAdditionalCardsDependantRoles = computed<boolean>(() => {
    const playersWithAdditionalCardsDependantRoles = getPlayersWithAnyRoleNameInCreateGameDto([...ADDITIONAL_CARDS_DEPENDANT_ROLES]);

    return playersWithAdditionalCardsDependantRoles.length > 0;
  });

  function setCreateGameDto(createGameDtoValue: CreateGameDto): void {
    createGameDto.value = CreateGameDto.create(createGameDtoValue);
    saveCreateGameOptionsDtoToLocalStorage();
  }

  function resetCreateGameDto(doesRetrieveLocalStorageValues = true): void {
    if (!doesRetrieveLocalStorageValues) {
      createGameDto.value = CreateGameDto.create(defaultCreateGameDto);
      saveCreateGameOptionsDtoToLocalStorage();

      return;
    }
    createGameDto.value = CreateGameDto.create({
      ...defaultCreateGameDto,
      options: createGameOptionsDtoFromLocalStorage.value,
    });
  }

  function resetCreateGameOptionsDto(): void {
    createGameDto.value.options = GameOptions.create(DEFAULT_GAME_OPTIONS);
    saveCreateGameOptionsDtoToLocalStorage();
  }

  function saveCreateGameOptionsDtoToLocalStorage(): void {
    createGameOptionsDtoFromLocalStorage.value = createGameDto.value.options;
  }

  function removeObsoleteAdditionalCardsFromCreateGameDto(): void {
    if (!createGameDto.value.additionalCards) {
      return;
    }
    const rolesInGame = new Set(createGameDto.value.players.map(player => player.role.name));
    createGameDto.value.additionalCards = createGameDto.value.additionalCards.filter(({ recipient }) => rolesInGame.has(recipient));
  }

  function addPlayerToCreateGameDto(player: CreateGamePlayerDto): void {
    createGameDto.value.players.push(CreateGamePlayerDto.create(player));
  }

  function updatePlayerInCreateGameDto(player: CreateGamePlayerDto): void {
    const playerIndex = createGameDto.value.players.findIndex(({ name }) => name === player.name);
    if (playerIndex !== -1) {
      createGameDto.value.players.splice(playerIndex, 1, CreateGamePlayerDto.create(player));
    }
  }

  function setPlayersToCreateGameDto(players: CreateGamePlayerDto[]): void {
    createGameDto.value.players = players.map(player => CreateGamePlayerDto.create(player));
  }

  function removePlayerFromCreateGameDto(playerName: string): void {
    const playerIndex = createGameDto.value.players.findIndex(player => player.name === playerName);
    if (playerIndex !== -1) {
      createGameDto.value.players = createGameDto.value.players.toSpliced(playerIndex, 1);
    }
  }

  function isRoleInCreateGameDto(roleName: RoleName): boolean {
    return createGameDto.value.players.some(player => player.role.name === roleName);
  }

  function getPlayersWithRoleNameInCreateGameDto(roleName: RoleName): CreateGamePlayerDto[] {
    return createGameDto.value.players.filter(player => player.role.name === roleName);
  }

  function getPlayersWithAnyRoleNameInCreateGameDto(roleNames: RoleName[]): CreateGamePlayerDto[] {
    return createGameDto.value.players.filter(player => player.role.name && roleNames.includes(player.role.name));
  }

  function isRoleMinReachedInCreateGameDto(roleName: RoleName): boolean {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);

    return !!role && (role.minInGame === undefined || playersWithRoleName.length >= role.minInGame);
  }

  function isRoleMaxReachedInCreateGameDto(roleName: RoleName): boolean {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);

    return !!role && playersWithRoleName.length >= role.maxInGame;
  }

  function getRoleLeftCountToReachMinInCreateGameDto(roleName: RoleName): number {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);
    if (role?.minInGame === undefined) {
      return 0;
    }
    const leftCount = role.minInGame - playersWithRoleName.length;

    return leftCount <= -1 ? 0 : leftCount;
  }

  function setAdditionalCardsForRecipientInCreateGameDto(additionalCards: CreateGameAdditionalCardDto[], recipient: GameAdditionalCardRecipientRoleName): void {
    if (!createGameDto.value.additionalCards) {
      createGameDto.value.additionalCards = additionalCards;
    }

    createGameDto.value.additionalCards = createGameDto.value.additionalCards.filter(card => card.recipient !== recipient);
    createGameDto.value.additionalCards.push(...additionalCards);
  }

  function getAdditionalCardsForRecipientInCreateGameDto(recipient: GameAdditionalCardRecipientRoleName): CreateGameAdditionalCardDto[] {
    return createGameDto.value.additionalCards?.filter(card => card.recipient === recipient) ?? [];
  }

  function getAdditionalCardsWithRoleNameInCreateGameDto(roleName: RoleName): CreateGameAdditionalCardDto[] {
    return createGameDto.value.additionalCards?.filter(card => card.roleName === roleName) ?? [];
  }
  return {
    createGameOptionsDtoFromLocalStorage,
    createGameDto,
    createGameOptionsDto,
    doesCreateGameDtoContainPositionDependantRoles,
    doesCreateGameDtoContainAdditionalCardsDependantRoles,
    setCreateGameDto,
    resetCreateGameDto,
    resetCreateGameOptionsDto,
    saveCreateGameOptionsDtoToLocalStorage,
    removeObsoleteAdditionalCardsFromCreateGameDto,
    addPlayerToCreateGameDto,
    updatePlayerInCreateGameDto,
    setPlayersToCreateGameDto,
    removePlayerFromCreateGameDto,
    isRoleInCreateGameDto,
    getPlayersWithRoleNameInCreateGameDto,
    getPlayersWithAnyRoleNameInCreateGameDto,
    isRoleMinReachedInCreateGameDto,
    isRoleMaxReachedInCreateGameDto,
    getRoleLeftCountToReachMinInCreateGameDto,
    setAdditionalCardsForRecipientInCreateGameDto,
    getAdditionalCardsForRecipientInCreateGameDto,
    getAdditionalCardsWithRoleNameInCreateGameDto,
  };
});

export { useCreateGameDtoStore };