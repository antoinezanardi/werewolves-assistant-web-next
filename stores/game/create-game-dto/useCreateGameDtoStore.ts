import { defineStore } from "pinia";
import { ref } from "vue";

import { CreateGamePlayerDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import { StoreIds } from "~/stores/enums/store.enum";

const useCreateGameDtoStore = defineStore(StoreIds.CREATE_GAME_DTO, () => {
  const createGameDto = ref<CreateGameDto>(CreateGameDto.create({ players: [] }));

  function setCreateGameDto(createGameDtoValue: CreateGameDto): void {
    createGameDto.value = CreateGameDto.create(createGameDtoValue);
  }

  function resetCreateGameDto(): void {
    createGameDto.value = CreateGameDto.create({ players: [] });
  }

  function addPlayerToCreateGameDto(player: CreateGamePlayerDto): void {
    createGameDto.value.players.push(CreateGamePlayerDto.create(player));
  }

  function updatePlayerInCreateGameDto(player: CreateGamePlayerDto): void {
    const playerIndex = createGameDto.value.players.findIndex(({ name }) => name === player.name);
    if (playerIndex !== -1) {
      createGameDto.value.players.splice(playerIndex, 1, CreateGamePlayerDto.create(player));
    }
  }

  function setPlayersToCreateGameDto(players: CreateGamePlayerDto[]): void {
    createGameDto.value.players = players.map(player => CreateGamePlayerDto.create(player));
  }

  function removePlayerFromCreateGameDto(playerName: string): void {
    const playerIndex = createGameDto.value.players.findIndex(player => player.name === playerName);
    if (playerIndex !== -1) {
      createGameDto.value.players = createGameDto.value.players.toSpliced(playerIndex, 1);
    }
  }
  return {
    createGameDto,
    setCreateGameDto,
    resetCreateGameDto,
    addPlayerToCreateGameDto,
    updatePlayerInCreateGameDto,
    setPlayersToCreateGameDto,
    removePlayerFromCreateGameDto,
  };
});

export { useCreateGameDtoStore };