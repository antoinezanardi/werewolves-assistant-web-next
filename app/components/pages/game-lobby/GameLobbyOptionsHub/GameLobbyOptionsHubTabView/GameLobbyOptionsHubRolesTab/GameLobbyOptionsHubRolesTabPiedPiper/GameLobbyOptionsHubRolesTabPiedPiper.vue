<template>
  <PrimeVueFieldset
    id="game-lobby-options-hub-roles-tab-pied-piper"
    :pt="{ 'legend': 'ml-4 !p-3 !px-6' }"
  >
    <template #legend>
      <GameOptionRoleLegend role-name="pied-piper"/>
    </template>

    <GameOptionInputGroup
      id="game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-input-group"
      does-have-bottom-divider
      :option-description="piedPiperCharmedPeopleCountPerNightDescription"
      option-icon="people-line"
      option-icon-class="text-pink-500"
      :option-label="$t('components.GameLobbyOptionsHubRolesTabPiedPiper.options.charmedPeopleCountPerNight.label')"
    >
      <PrimeVueFloatLabel>
        <label
          id="game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-label"
          for="game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-input"
        >
          {{ $t("components.GameLobbyOptionsHubRolesTabPiedPiper.options.charmedPeopleCountPerNight.label") }}
        </label>

        <PrimeVueInputNumber
          id="game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-input"
          v-model="piedPiperCharmedPeopleCountPerNightValue"
          class="w-full"
          :max="5"
          :min="1"
        />

        <PrimeVueSlider
          id="game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-slider"
          v-model="piedPiperCharmedPeopleCountPerNightValue"
          class="mt-4 w-full"
          :max="5"
          :min="1"
          :step="1"
        />
      </PrimeVueFloatLabel>
    </GameOptionInputGroup>

    <GameOptionInputGroup
      id="game-lobby-options-hub-roles-tab-pied-piper-is-powerless-on-werewolves-side-input-group"
      does-have-bottom-divider
      :option-description="isPiedPiperPowerlessOnWerewolvesSideDescription"
      option-icon="ban"
      option-icon-class="text-red-500"
      :option-label="$t('components.GameLobbyOptionsHubRolesTabPiedPiper.options.isPowerlessOnWerewolvesSide.label')"
    >
      <AffirmativeToggleButton
        id="game-lobby-options-hub-roles-tab-pied-piper-is-powerless-on-werewolves-side-input"
        v-model="isPiedPiperPowerlessOnWerewolvesSideValue"
        class="w-full"
      />
    </GameOptionInputGroup>

    <GameOptionInputGroup
      id="game-lobby-options-hub-roles-tab-pied-piper-are-charmed-people-revealed-input-group"
      :option-description="areCharmedPeopleByPiedPiperRevealedDescription"
      option-icon="bullhorn"
      option-icon-class="text-blue-500"
      :option-label="$t('components.GameLobbyOptionsHubRolesTabPiedPiper.options.areCharmedPeopleRevealed.label')"
    >
      <AffirmativeToggleButton
        id="game-lobby-options-hub-roles-tab-pied-piper-are-charmed-people-revealed-input"
        v-model="areCharmedPeopleByPiedPiperRevealedValue"
        class="w-full"
      />
    </GameOptionInputGroup>
  </PrimeVueFieldset>
</template>

<script setup lang="ts">
import { storeToRefs } from "pinia";
import AffirmativeToggleButton from "~/components/shared/buttons/AffirmativeToggleButton/AffirmativeToggleButton.vue";
import GameOptionInputGroup from "~/components/shared/game/game-options/GameOptionInputGroup/GameOptionInputGroup.vue";
import GameOptionRoleLegend from "~/components/shared/game/game-options/GameOptionRoleLegend/GameOptionRoleLegend.vue";
import { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import { useStrings } from "~/composables/misc/useStrings";
import { useCreateGameDtoStore } from "~/stores/game/create-game-dto/useCreateGameDtoStore";

const createGameDtoStore = useCreateGameDtoStore();
const { setCreateGameDto } = createGameDtoStore;
const { createGameDto } = storeToRefs(createGameDtoStore);

const { t } = useI18n();

const { convertBooleanAsAffirmativeString } = useStrings();

const piedPiperCharmedPeopleCountPerNightValue = computed<number>({
  get: () => createGameDto.value.options.roles.piedPiper.charmedPeopleCountPerNight,
  set: (value: number | null) => {
    if (value === null) {
      return;
    }
    const localCreateGameDto = CreateGameDto.create(createGameDto.value);
    localCreateGameDto.options.roles.piedPiper.charmedPeopleCountPerNight = value;
    setCreateGameDto(localCreateGameDto);
  },
});

const isPiedPiperPowerlessOnWerewolvesSideValue = computed<boolean>({
  get: () => createGameDto.value.options.roles.piedPiper.isPowerlessOnWerewolvesSide,
  set: (value: boolean) => {
    const localCreateGameDto = CreateGameDto.create(createGameDto.value);
    localCreateGameDto.options.roles.piedPiper.isPowerlessOnWerewolvesSide = value;
    setCreateGameDto(localCreateGameDto);
  },
});

const areCharmedPeopleByPiedPiperRevealedValue = computed<boolean>({
  get: () => createGameDto.value.options.roles.piedPiper.areCharmedPeopleRevealed,
  set: (value: boolean) => {
    const localCreateGameDto = CreateGameDto.create(createGameDto.value);
    localCreateGameDto.options.roles.piedPiper.areCharmedPeopleRevealed = value;
    setCreateGameDto(localCreateGameDto);
  },
});

const piedPiperCharmedPeopleCountPerNightDescription = computed<string>(() => {
  const tKey = "components.GameLobbyOptionsHubRolesTabPiedPiper.options.charmedPeopleCountPerNight.description";

  return t(tKey, { count: piedPiperCharmedPeopleCountPerNightValue.value });
});

const isPiedPiperPowerlessOnWerewolvesSideDescription = computed<string>(() => {
  const booleanAsAffirmative = convertBooleanAsAffirmativeString(isPiedPiperPowerlessOnWerewolvesSideValue.value);

  return t(`components.GameLobbyOptionsHubRolesTabPiedPiper.options.isPowerlessOnWerewolvesSide.descriptions.${booleanAsAffirmative}`);
});

const areCharmedPeopleByPiedPiperRevealedDescription = computed<string>(() => {
  const booleanAsAffirmative = convertBooleanAsAffirmativeString(areCharmedPeopleByPiedPiperRevealedValue.value);

  return t(`components.GameLobbyOptionsHubRolesTabPiedPiper.options.areCharmedPeopleRevealed.descriptions.${booleanAsAffirmative}`);
});
</script>