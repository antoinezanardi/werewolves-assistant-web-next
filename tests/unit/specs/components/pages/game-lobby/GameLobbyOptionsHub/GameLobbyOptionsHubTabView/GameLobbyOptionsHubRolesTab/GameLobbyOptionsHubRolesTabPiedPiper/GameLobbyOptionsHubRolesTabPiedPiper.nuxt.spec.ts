import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";
import type InputNumber from "primevue/inputnumber";
import type Slider from "primevue/slider";
import GameLobbyOptionsHubRolesTabPiedPiper from "~/components/pages/game-lobby/GameLobbyOptionsHub/GameLobbyOptionsHubTabView/GameLobbyOptionsHubRolesTab/GameLobbyOptionsHubRolesTabPiedPiper/GameLobbyOptionsHubRolesTabPiedPiper.vue";
import type AffirmativeToggleButton from "~/components/shared/buttons/AffirmativeToggleButton/AffirmativeToggleButton.vue";
import type GameOptionInputGroup from "~/components/shared/game/game-options/GameOptionInputGroup/GameOptionInputGroup.vue";
import { useCreateGameDtoStore } from "~/stores/game/create-game-dto/useCreateGameDtoStore";
import { createFakeCreateGameDto } from "@tests/unit/utils/factories/composables/api/game/dto/create-game/create-game.dto.factory";

import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import type { VueVm } from "@tests/unit/utils/types/vue-test-utils.types";

describe("Game Lobby Options Hub Roles Tab Pied Piper Component", () => {
  let wrapper: ReturnType<typeof mount<typeof GameLobbyOptionsHubRolesTabPiedPiper>>;

  async function mountGameLobbyOptionsHubRolesTabPiedPiperComponent(options: ComponentMountingOptions<typeof GameLobbyOptionsHubRolesTabPiedPiper> = {}):
  Promise<ReturnType<typeof mount<typeof GameLobbyOptionsHubRolesTabPiedPiper>>> {
    return mountSuspendedComponent(GameLobbyOptionsHubRolesTabPiedPiper, {
      global: {
        stubs: {
          Fieldset: false,
          GameOptionInputGroup: false,
          FloatLabel: false,
        },
      },
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameLobbyOptionsHubRolesTabPiedPiperComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Charmed People count per night option", () => {
    it("should translate option label when rendered.", () => {
      const gameOptionInputGroup = wrapper.findComponent<typeof GameOptionInputGroup>("#game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-input-group");

      expect(gameOptionInputGroup.props("optionLabel")).toBe("Charmed people count per night");
    });

    it("should translate option description with selected value when rendered.", () => {
      const gameOptionInputGroup = wrapper.findComponent<typeof GameOptionInputGroup>("#game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-input-group");
      const expectedOptionDescription = "composables.useGameOptionsTexts.roles.piedPiper.charmedPeopleCountPerNight, {\"charmedPeopleCountPerNight\":2}, 2";

      expect(gameOptionInputGroup.props("optionDescription")).toBe(expectedOptionDescription);
    });

    it("should translate option float label when rendered.", () => {
      const floatLabel = wrapper.find<HTMLLabelElement>("#game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-label");

      expect(floatLabel.text()).toBe("Charmed people count per night");
    });

    it("should update the create game dto store when the option is changed by the input number.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      const inputId = "#game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-input";
      const charmedPeopleCountPerNightInputNumber = wrapper.findComponent<typeof InputNumber>(inputId);
      (charmedPeopleCountPerNightInputNumber.vm as VueVm).$emit("update:modelValue", 3);
      await wrapper.vm.$nextTick();
      const expectedCreateGameDto = createFakeCreateGameDto(createGameDtoStore.createGameDto);
      expectedCreateGameDto.options.roles.piedPiper.charmedPeopleCountPerNight = 3;

      expect(createGameDtoStore.setCreateGameDto).toHaveBeenCalledExactlyOnceWith(expectedCreateGameDto);
    });

    it("should update the create game dto store when the option is changed by the slider.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      const inputId = "#game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-slider";
      const charmedPeopleCountPerNightInputNumber = wrapper.findComponent<typeof Slider>(inputId);
      (charmedPeopleCountPerNightInputNumber.vm as VueVm).$emit("update:modelValue", 4);
      await wrapper.vm.$nextTick();
      const expectedCreateGameDto = createFakeCreateGameDto(createGameDtoStore.createGameDto);
      expectedCreateGameDto.options.roles.piedPiper.charmedPeopleCountPerNight = 4;

      expect(createGameDtoStore.setCreateGameDto).toHaveBeenCalledExactlyOnceWith(expectedCreateGameDto);
    });

    it("should not update the create game dto store when the option is changed by the input number and the value is null.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      const inputId = "#game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-input";
      const charmedPeopleCountPerNightInputNumber = wrapper.findComponent<typeof InputNumber>(inputId);
      (charmedPeopleCountPerNightInputNumber.vm as VueVm).$emit("update:modelValue", null);
      await wrapper.vm.$nextTick();

      expect(createGameDtoStore.setCreateGameDto).not.toHaveBeenCalled();
    });
  });

  describe("Is powerless on Werewolves Side", () => {
    it("should translate option label when rendered.", () => {
      const gameOptionInputGroup = wrapper.findComponent<typeof GameOptionInputGroup>("#game-lobby-options-hub-roles-tab-pied-piper-is-powerless-on-werewolves-side-input-group");

      expect(gameOptionInputGroup.props("optionLabel")).toBe("The Pied Piper is powerless on the Werewolves side");
    });

    it("should translate option description when the option is activated.", () => {
      const gameOptionInputGroup = wrapper.findComponent<typeof GameOptionInputGroup>("#game-lobby-options-hub-roles-tab-pied-piper-is-powerless-on-werewolves-side-input-group");

      expect(gameOptionInputGroup.props("optionDescription")).toBe("composables.useGameOptionsTexts.roles.piedPiper.isPowerlessOnWerewolvesSide.yes");
    });

    it("should translate option description when the option is deactivated.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      createGameDtoStore.createGameDto.options.roles.piedPiper.isPowerlessOnWerewolvesSide = false;
      const gameOptionInputGroup = wrapper.findComponent<typeof GameOptionInputGroup>("#game-lobby-options-hub-roles-tab-pied-piper-is-powerless-on-werewolves-side-input-group");
      await nextTick();

      expect(gameOptionInputGroup.props("optionDescription")).toBe("composables.useGameOptionsTexts.roles.piedPiper.isPowerlessOnWerewolvesSide.no");
    });

    it("should update the create game dto store when the option is toggled.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      const toggleButton = wrapper.findComponent<typeof AffirmativeToggleButton>("#game-lobby-options-hub-roles-tab-pied-piper-is-powerless-on-werewolves-side-input");
      (toggleButton.vm as VueVm).$emit("update:modelValue", false);
      await nextTick();
      const expectedCreateGameDto = createFakeCreateGameDto(createGameDtoStore.createGameDto);
      expectedCreateGameDto.options.roles.piedPiper.isPowerlessOnWerewolvesSide = false;

      expect(createGameDtoStore.setCreateGameDto).toHaveBeenCalledExactlyOnceWith(expectedCreateGameDto);
    });
  });

  describe("Are charmed people revealed", () => {
    it("should translate option label when rendered.", () => {
      const gameOptionInputGroup = wrapper.findComponent<typeof GameOptionInputGroup>("#game-lobby-options-hub-roles-tab-pied-piper-are-charmed-people-revealed-input-group");

      expect(gameOptionInputGroup.props("optionLabel")).toBe("Charmed people are revealed");
    });

    it("should translate option description when the option is activated.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      createGameDtoStore.createGameDto.options.roles.piedPiper.areCharmedPeopleRevealed = true;
      const gameOptionInputGroup = wrapper.findComponent<typeof GameOptionInputGroup>("#game-lobby-options-hub-roles-tab-pied-piper-are-charmed-people-revealed-input-group");
      await nextTick();

      expect(gameOptionInputGroup.props("optionDescription")).toBe("composables.useGameOptionsTexts.roles.piedPiper.areCharmedPeopleRevealed.yes");
    });

    it("should translate option description when the option is deactivated.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      createGameDtoStore.createGameDto.options.roles.piedPiper.areCharmedPeopleRevealed = false;
      const gameOptionInputGroup = wrapper.findComponent<typeof GameOptionInputGroup>("#game-lobby-options-hub-roles-tab-pied-piper-are-charmed-people-revealed-input-group");
      await nextTick();

      expect(gameOptionInputGroup.props("optionDescription")).toBe("composables.useGameOptionsTexts.roles.piedPiper.areCharmedPeopleRevealed.no");
    });

    it("should update the create game dto store when the option is toggled.", async() => {
      const createGameDtoStore = useCreateGameDtoStore();
      const toggleButton = wrapper.findComponent<typeof AffirmativeToggleButton>("#game-lobby-options-hub-roles-tab-pied-piper-are-charmed-people-revealed-input");
      (toggleButton.vm as VueVm).$emit("update:modelValue", false);
      await nextTick();
      const expectedCreateGameDto = createFakeCreateGameDto(createGameDtoStore.createGameDto);
      expectedCreateGameDto.options.roles.piedPiper.areCharmedPeopleRevealed = false;

      expect(createGameDtoStore.setCreateGameDto).toHaveBeenCalledExactlyOnceWith(expectedCreateGameDto);
    });
  });
});