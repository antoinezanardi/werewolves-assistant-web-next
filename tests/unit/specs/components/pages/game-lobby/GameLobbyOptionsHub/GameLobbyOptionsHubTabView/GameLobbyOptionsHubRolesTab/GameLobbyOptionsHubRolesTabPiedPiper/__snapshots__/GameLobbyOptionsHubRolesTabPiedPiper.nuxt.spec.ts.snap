// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Lobby Options Hub Roles Tab Pied Piper Component > should match snapshot when rendered. 1`] = `
"<fieldset class="p-fieldset p-component" id="game-lobby-options-hub-roles-tab-pied-piper" data-pc-name="fieldset" data-pc-section="root" pc2="">
  <legend class="p-fieldset-legend ml-4 !p-3 !px-6" data-pc-section="legend">
    <game-option-role-legend-stub rolename="pied-piper"></game-option-role-legend-stub>
  </legend>
  <transition-stub name="p-toggleable-content" appear="false" persisted="false" css="true">
    <div id="game-lobby-options-hub-roles-tab-pied-piper_content" class="p-fieldset-content-container" role="region" aria-labelledby="game-lobby-options-hub-roles-tab-pied-piper_header" data-pc-section="contentcontainer">
      <div class="p-fieldset-content" data-pc-section="content">
        <div id="game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-input-group">
          <div id="game-option-input-group" class="flex gap-2 items-center justify-between">
            <h3 id="game-option-label" class="w-8/12">
              <font-awesome-icon-stub icon="people-line" border="false" fixedwidth="false" flip="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" bounce="false" shake="false" beat="false" fade="false" beatfade="false" flash="false" spinpulse="false" spinreverse="false" id="game-option-label-icon" class="me-2 text-pink-500"></font-awesome-icon-stub><span>Charmed people count per night</span>
            </h3>
            <div id="game-option-input" class="w-4/12"><span class="p-floatlabel" data-pc-name="floatlabel" data-pc-section="root" pc1=""><label id="game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-label" for="game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-input">Charmed people count per night</label><input-number-stub id="game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-input" modelvalue="2" class="w-full" max="5" min="1"></input-number-stub><slider-stub id="game-lobby-options-hub-roles-tab-pied-piper-charmed-people-count-per-night-slider" modelvalue="2" class="mt-4 w-full" max="5" min="1" step="1"></slider-stub></span></div>
          </div>
          <p id="game-option-description" class="italic mt-3 text-gray-400">components.GameLobbyOptionsHubRolesTabPiedPiper.options.charmedPeopleCountPerNight.description, {"count":2}</p>
          <divider-stub id="game-option-bottom-divider" class="!my-4"></divider-stub>
        </div>
        <div id="game-lobby-options-hub-roles-tab-pied-piper-is-powerless-on-werewolves-side-input-group">
          <div id="game-option-input-group" class="flex gap-2 items-center justify-between">
            <h3 id="game-option-label" class="w-8/12">
              <font-awesome-icon-stub icon="ban" border="false" fixedwidth="false" flip="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" bounce="false" shake="false" beat="false" fade="false" beatfade="false" flash="false" spinpulse="false" spinreverse="false" id="game-option-label-icon" class="me-2 text-red-500"></font-awesome-icon-stub><span>The Pied Piper is powerless on the Werewolves side</span>
            </h3>
            <div id="game-option-input" class="w-4/12">
              <affirmative-toggle-button-stub modelvalue="true" id="game-lobby-options-hub-roles-tab-pied-piper-is-powerless-on-werewolves-side-input" class="w-full"></affirmative-toggle-button-stub>
            </div>
          </div>
          <p id="game-option-description" class="italic mt-3 text-gray-400">components.GameLobbyOptionsHubRolesTabPiedPiper.options.isPowerlessOnWerewolvesSide.descriptions.yes</p>
          <divider-stub id="game-option-bottom-divider" class="!my-4"></divider-stub>
        </div>
        <div id="game-lobby-options-hub-roles-tab-pied-piper-are-charmed-people-revealed-input-group">
          <div id="game-option-input-group" class="flex gap-2 items-center justify-between">
            <h3 id="game-option-label" class="w-8/12">
              <font-awesome-icon-stub icon="bullhorn" border="false" fixedwidth="false" flip="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" bounce="false" shake="false" beat="false" fade="false" beatfade="false" flash="false" spinpulse="false" spinreverse="false" id="game-option-label-icon" class="me-2 text-blue-500"></font-awesome-icon-stub><span>Charmed people are revealed</span>
            </h3>
            <div id="game-option-input" class="w-4/12">
              <affirmative-toggle-button-stub modelvalue="false" id="game-lobby-options-hub-roles-tab-pied-piper-are-charmed-people-revealed-input" class="w-full"></affirmative-toggle-button-stub>
            </div>
          </div>
          <p id="game-option-description" class="italic mt-3 text-gray-400">components.GameLobbyOptionsHubRolesTabPiedPiper.options.areCharmedPeopleRevealed.descriptions.no</p>
          <!--v-if-->
        </div>
      </div>
    </div>
  </transition-stub>
</fieldset>"
`;
